### ISSUE TITLE
Remove user message from message list when `USER` message is `None`

### ISSUE THREAD

**Issue by off6atomic** [Mirascope/mirascope#265]
### Description

Sometimes I want to remove user message from `messages()` list because I want to force the chatbot to look at history only. This is particularly useful when tool output is the last message in the chat history. I want to force the chatbot to think about tool output without thinking about extra user message.

The below code demonstrates current behavior which still shows user message on the list even after setting it to `None`.

```py
class Forecast(OpenAICall):
    prompt_template = """
    MESSAGES: {history}
    USER: {question}
    """

    question: str | None
    history: list[ChatCompletionMessageParam] = []
    call_params = OpenAICallParams(model="gpt-4-turbo")


print(Forecast(
    question=None,
    history=[{"role": "user", "content": "What is the weather in Tokyo?"}],
).messages())
# >> [{'content': 'What is the weather in Tokyo?', 'role': 'user'}, {'role': 'user', 'content': 'None'}]
# I don't want to see {'role': 'user', 'content': 'None'}
```

Is this feature desirable?

**Comment by willbakst** [comment]
Can you elaborate on instances in which you expect the user message to be `None` and it doesn't just make sense for you to handle that case on your end?

**Comment by off6atomic** [comment]
For example, let's say we are just trying to make a chatbot that answers about the weather like in #261

```py
from typing import Literal

from openai.types.chat import ChatCompletionMessageParam

from mirascope.openai import OpenAICall, OpenAICallParams


def get_current_weather(
    location: str, unit: Literal["celsius", "fahrenheit"] = "fahrenheit"
):
    """Get the current weather in a given location."""
    if "tokyo" in location.lower():
        return f"It is 10 degrees {unit} in Tokyo, Japan"
    elif "san francisco" in location.lower():
        return f"It is 72 degrees {unit} in San Francisco, CA"
    elif "paris" in location.lower():
        return f"It is 22 degress {unit} in Paris, France"
    else:
        return f"I'm not sure what the weather is like in {location}"


class Forecast(OpenAICall):
    prompt_template = """
    MESSAGES: {history}
    USER: {question}
    """

    question: str
    history: list[ChatCompletionMessageParam] = []
    call_params = OpenAICallParams(model="gpt-4-turbo", tools=[get_current_weather])


# Make the first call to the LLM
forecast = Forecast(question="What's the weather in Tokyo Japan?")
response = forecast.call()
forecast.history += [
    {"role": "user", "content": forecast.question},
    response.message.model_dump(),
]

tool = response.tool
if tool:
    print("Tool arguments:", tool.args)
    # > {'location': 'Tokyo, Japan', 'unit': 'fahrenheit'}
    output = tool.fn(**tool.args)
    print("Tool output:", output)
    # > It is 10 degrees fahrenheit in Tokyo, Japan

    # reinsert the tool call into the chat messages through history
    # NOTE: this should be more convenient, e.g. `tool.message_param`
    forecast.history += [
        {
            "role": "tool",
            "content": output,
            "tool_call_id": tool.tool_call.id,
            "name": tool.__class__.__name__,
        },
    ]
else:
    print(response.content)  # if no tool, print the content of the response

# Call the LLM again with the history including the tool call
forecast.question = "Is that cold or hot?"
response = forecast.call()
print("After Tools Response:", response.content)
```

This is the output:
```txt
Tool arguments: {'location': 'Tokyo, Japan', 'unit': 'fahrenheit'}
Tool output: It is 10 degrees fahrenheit in Tokyo, Japan
After Tools Response: 10 degrees Fahrenheit is very cold. This temperature is well below freezing and typically occurs during winter months in colder regions.
```

This seems to be OK. But what if I don't want to ask the bot whether it's hard or cold?
What if I just want it to look at the tool output and answer the original question: "What's the weather in Tokyo Japan?"

How can I do that?

A naive idea is to set last user question to be empty instead of setting it to "Is that cold or hot?".
But this is a hack because you still have the user message in the list.

If you print `forecast.messages()`, you will see the following output:
```py
[{'role': 'user', 'content': "What's the weather in Tokyo Japan?"},
 {'content': None,
  'role': 'assistant',
  'function_call': None,
  'tool_calls': [{'id': 'call_Fp7sUiUdNa45fpwEeDuafSzg',
    'function': {'arguments': '{"location":"Tokyo, Japan"}',
     'name': 'GetCurrentWeather'},
    'type': 'function'}]},
 {'role': 'tool',
  'content': 'It is 10 degrees fahrenheit in Tokyo, Japan',
  'tool_call_id': 'call_Fp7sUiUdNa45fpwEeDuafSzg',
  'name': 'GetCurrentWeather'},
 {'role': 'user', 'content': ''}] # notice that this line exists even though it has no content. I just want to remove it
```

How can I remove that last message in the list?

A workaround that I thought of is to create a new Forecast class that doesn't have question input like this:
```py
class ForecastWithoutQuestion(OpenAICall):
    prompt_template = """
    MESSAGES: {history}
    """

    history: list[ChatCompletionMessageParam] = []
    call_params = OpenAICallParams(model="gpt-4-turbo", tools=[get_current_weather])
```

Not sure if this is what we should do when the input can be empty though.

**Comment by off6atomic** [comment]
As a side note, this also raises interesting question about the current behavior of prompt template which is quite static. What if users want to change prompt based on values of inputs?

For example, let's say you are writing a prompt for writing an article in a style of a certain writer, but sometimes you don't want to specify writer, what you have to do right now is to create 2 classes:
```py
class WriteArticle(OpenAICall):
    prompt_template = "Write an article about {topic}"

    topic: str
    call_params = OpenAICallParams(model="gpt-4-turbo")


class WriteArticleWithStyle(OpenAICall):
    prompt_template = "Write an article about {topic} in the style of {author}."

    topic: str
    author: str
    call_params = OpenAICallParams(model="gpt-4-turbo")
```

Do you like the current behavior as it is?

**Comment by willbakst** [comment]
First, to answer your latter question about "static" prompt templating.

Part of me thinks these should in fact be two different classes, especially considering that down the line you may further prompt engineer the prompt that has the author vs. doesn't have the author to get better responses in each case.

However, we don't want to force this on users of course. You can do something like the following to make it more dynamic in the current version:

```python
class WriteArticle(OpenAICall):
    prompt_template = "Write 100 word article about {topic}{optional_author_style}"

    topic: str
    author: Optional[str]

    @property
    def optional_author_style(self) -> str:
        return "" if self.author is None else f" in the style of {self.author}"
```

Would this suffice for this case?

**Comment by offchan42** [comment]
Yes it suffices. Can I make prompt_template as property like in the example I provided?
Because that's another alternative to creating multiple classes especially when all the classes might have a lot of the same inputs.

**Comment by willbakst** [comment]
Right now `prompt_template` is a `ClassVar`, so you can't write it as a property.

I'm not sure exactly what this would look like, but feel free to post a separate feature request issue for something like this so we can further flesh out what that might look like if we want to include it.

**Comment by willbakst** [comment]
Fix for this is (finally after GitHub actions was failing me) released in v0.13.5 🎉
